packages/@lwc/integration-karma/test/mixed-shadow-mode/contextualComposedPath/index.spec.js

```diff
@@ -0,0 +1,48 @@
+import { createElement } from 'lwc';
+import Parent from 'x/parent';
+
+// https://github.com/salesforce/lwc/issues/4011
+describe('native inside synthetic', () => {
+    it('should have shadowed elements in composed path when inside shadow boundary', () => {
+        const element = createElement('x-parent', { is: Parent });
+        element.setAttribute('x-parent');
+        document.body.appendChild(element);
+
+        return Promise.resolve().then(() => {
+            element.click();
+
+            expect(element.childComposedPath).toEqual([
+                jasmine.any(HTMLElement),
+                jasmine.any(HTMLElement),
+                jasmine.any(HTMLElement),
+                jasmine.any(Object), // #document-fragment for closed shadow root
+                element.child,
+                element.shadowRoot,
+                element,
+                document.body,
+                document.documentElement,
+                document,
+                window,
+            ]);
```

Comment:
You can use `extractDataIds` here to make this a bit more readable.

Reply:
We were using that initially but it doesn't work with closed shadow roots.